pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,378 @@
+		// can happen if we discover that the prefix is also associated with
+		// other labels besides kube-apiserver, i.e. if the kube-apiserver is
+		// not running outside of the cluster.
+		return injectLabelsForCIDR(prefix, lbls)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
+	defer cancel()
+	return IdentityAllocator.AllocateIdentity(ctx, lbls, false)
+}
+
+// injectLabelsForCIDR will allocate a CIDR identity for the given prefix. The
+// release of the identity must be managed by the caller.
+func injectLabelsForCIDR(prefix string, lbls labels.Labels) (*identity.Identity, bool, error) {
+	if !strings.Contains(prefix, "/") {
+		prefix = prefix + "/32"
+	}
+	_, cidr, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, false, err
+	}
+
+	allLbls := cidrlabels.GetCIDRLabels(cidr)
+	allLbls.MergeLabels(lbls)
+
+	log.WithFields(logrus.Fields{
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the IdentityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	matching := make([]string, 0)
+	for prefix, lbls := range IdentityMetadata {
+		if bytes.Contains(lbls.SortedList(), filter.SortedList()) {
+			matching = append(matching, prefix)
+		}
+	}
+	return matching
+}
+
+// RemoveAllPrefixesWithLabels wraps RemoveLabels to provide a convenient
+// method for the caller to remove all given prefixes at once. This function
+// will trigger policy update and recalculation if necessary on behalf of the
+// caller if any changes to the kube-apiserver were detected.
+//
+// Identities allocated by InjectLabels() may be released by RemoveLabels().
+//
+// A prefix will only be removed from the IDMD if the set of labels becomes
+// empty.
+func RemoveAllPrefixesWithLabels(
+	m map[string]labels.Labels,
+	src source.Source,
+	updater identityUpdater,
+	triggerer policyTriggerer,
+) {
+	var (
+		hasKubeAPIServer bool
+
+		idsToPropagate = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range m {
+		id, exists := IPIdentityCache.LookupByIP(prefix)
+		if id.ID <= identity.IdentityUnknown || !exists {
+			continue
+		}
+		if lbls.Has(labels.LabelKubeAPIServer[labels.IDNameKubeAPIServer]) {
+			hasKubeAPIServer = true
+		}
+		// Insert to propagate the updated set of labels after removal.
+		idsToPropagate[id.ID] = RemoveLabels(prefix, lbls, src).LabelArray()
+	}
+	if hasKubeAPIServer {
+		if updater == nil || triggerer == nil {
+			log.Warn("Unable to trigger policy updates after removing from ipcache because policy subsystem is not yet initialized.")
+			return
+		}
```

Comment:
This looks like more defensive coding that we should be able to remove with some better structuring of the codebase.

Reply:
Some of these were removed in the final version in https://github.com/cilium/cilium/pull/17823.